### DISCUSSION TITLE
Wechselrichter Daten in Systemvariable in Homematic

### DISCUSSION THREAD

**Discussion by Sven4597** [SukramJ/aiohomematic#621] [Q&A]
Hallo,

Ich würde gerne die Daten von meinem Wechselrichter den ich über Home Assistant auslese in eine Systemvariable von Homematic schreiben.

Mein Hauptsystem ist Homematic und habe auf einer extra Raspi das Home Assistant laufen.

zu Beispiel
Home Assistant: Lifetime solar energy 1000 KW
in Homematic Systemvariable: Lifetime solar energy
Die Werte sollen aktualisiert werden sobald sich was in Home Assistant ändert.

Könnt Ihr mir da bei der Umsetzung helfen? Oder ist es überhaupt möglich?

Ich kenne mich leider nicht so gut aus in Python.

Danke im vorraus
Grüße Sven

**Comment by emufan** [comment]
Da sollte doch eine extended SV reichen/passend sein, oder?

https://github.com/danielperna84/custom_homematic#system-variables

Oder ansonsten

https://github.com/danielperna84/custom_homematic#homematicip_localset_variable_value

Unabhängig davon, dass ich persönlich alles in HA verlagern würde als andersherum sogar noch HM mit Daten zu füttern.

**Reply by Sven4597** [reply]
Hallo emufan,
ich bin leider immer noch nicht weiter gekommen.
Könntest du mir ein Beispiel machen wie das dann auszusehen hat?

Gruß Sven

**Reply by emufan** [reply]
Da ich keine SVs nutze, kann ich Dir keine eigenen Beispiele senden. Aber ich würde es an Deiner Stelle einfach mal so ausprobieren wie oben in der Doku beschrieben. Das sollte doch klappen. Oder was genau klappt nicht, bzw. was genau hast Du schon (erfolglos?) probiert?

**Reply by Jonvelle53** [reply]
Hallo EMUFan,

ich habe so gut wie gar keine ahnung. Ich habe versucht die Zeilen von Sven bei mir nachzuvollziehen, da ich gern ein paar Bodenfeuchte Werte in Homematic SVariable schreiben würde. Die werden dann in der @Home App neben der Sprenger Aktivität mit angezeigt und zur Steuerung genutzt.
Ich habe ja schon Probleme die Zeilen in eine entsprechende Syntax zu bringen. Klappt nur wenn ich es versuche - Error und Trial.

So sieht meine Automation aus:
- id: '1691321639718'
  alias: Bodenfeuchte Werte in%
  description: übertragen
  trigger:
  - platform: state
    entity_id: sensor.gw2000a_v2_2_3_soil_moisture_1
  condition: []
  action:
  - service: homematicip_local.set_variable_value
    data: sensor.gw2000a_v2_2_3_soil_moisture_1 
    entry_id: ca0c16aXXXXXXXXXXd22f4bba694699a
    name: VH_Boden 
      value: > 
      {{ states('sensor.gw2000a_v2_2_3_soil_moisture_1') | float | 
      round(1) }}
  mode: single

Fehlermeldung vom Homeassistant:
Automation with alias 'Bodenfeuchte Werte in%' could not be validated and has been disabled: extra keys not allowed @ data['action'][0]['entry_id']. Got 'XXXXXXX303e0d70d0d22f4bba694699a' extra keys not allowed @ data['action'][0]['name']. Got "VH_Boden value:> {{ states('sensor.gw2000a_v2_2_3_soil_moisture_1') | float | round(1) }}"

Ich hatte die o.g. Zeilen in die config/automations.yaml datei geschrieben.

Wie gesagt, ich bin weder Programmierer noch wirklich ein Hardwarespezialist. Bin Nachrichtentechniker im Ruhestand seit 5 Jahren.
Soll heißen, dass ich das Meiste mir hier erarbeiten muss, und dann noch in einer mir nicht geläufigen englischen Sprache  :-(
Aber dafür gibts zum Glück ja Tools.

Würde mich freuen, wenn Du mir oder gern natürlich auch andere hier helfen könnten?

Mit besten Grüßen
Jonvelle

**Reply by Sven4597** [reply]
Hallo,

bitte versuche mal folgendes in einer neuen Automation als YAML einzufügen.
Die Großgeschriebenen bitte Löschen und durch die von dir ersetzen

[Neues Textdokument.txt](https://github.com/danielperna84/hahomematic/files/12269752/Neues.Textdokument.txt)

**Reply by Jonvelle53** [reply]
Hallo,

wo genau muss ich die Automation einfügen? config/automations.yaml ? Oder über Menue Punkt Automationen mit Yaml erzeugen?

**Reply by Sven4597** [reply]
Automationen mit Yaml erzeugen

**Reply by Sven4597** [reply]
![Unbenannt](https://github.com/danielperna84/hahomematic/assets/47686540/3676dec6-c9d5-48e7-984b-66eebad9fda3)

**Reply by Jonvelle53** [reply]
![Unbenannt](https://github.com/danielperna84/hahomematic/assets/124449225/15e95e68-847e-43f3-91ec-4b596e281dde)

**Reply by Jonvelle53** [reply]
Irgendwas denke ich falsch, wenn ich jetzt mit dem editor in config/automations schaue, steht nur eine leere Automation im File.

**Reply by Sven4597** [reply]
[Neues Textdokument.txt](https://github.com/danielperna84/hahomematic/files/12269874/Neues.Textdokument.txt)

versuche das mal in die Automations.yaml einzutragen und neustarten

**Reply by Jonvelle53** [reply]
can not read an implicit mapping pair; a colon is missed (726:77)

 723 |  ... 
 724 |  ... 
 725 |  ... 
 726 |  ... ture_1') | float | round(1) }}
------------------------------------------^
 727 |  ...

**Reply by Jonvelle53** [reply]
Ich sehe außerdem in der automations yaml Datei immer vor jeder Automation eine ID:XXXX
Kann man da die nummer von der mit Oberflächer erzeugten Automation eingeben?

**Reply by Sven4597** [reply]
Die ID musst du von dir übernehmen

**Reply by Jonvelle53** [reply]
![Yaml Fehler -kein Start möglich](https://github.com/danielperna84/hahomematic/assets/124449225/6de74965-3efb-406d-9141-33761ebd54b8)

**Reply by Jonvelle53** [reply]
Die "id" für den Eintrag einer Automation in der automations.yaml datei wird ja automatisch erzeugt, wenn ich imAutomationen Editor eine Automation neu erzeuge. Dein mir vorgeschlagener Yaml-Code einer Automation hat ja keine ID bzw. er bekommt keine wenn ich diesen Code per Fileeditor in die automations.yaml datei einfüge!

**Reply by Jonvelle53** [reply]
ok., habe nochmal gelöscht und direkt in die Yaml Automations Datei Deinen Vorschlag gespeichert.

Fehlermeldung wenn ich in den Entwicklerwewrkzeugen HA neu starten möchte:

Konfiguration fehlerhaft!
 Error loading /config/configuration.yaml: invalid key: "{"states('sensor.gw2000a_v2_2_3_soil_moisture_1') | float | round(1)": None}" in "/config/automations.yaml", line 727, column 0

Der Eintrag in der Automations Datei von Dir sieht so aus:

id: 'XXXXXXXXXX'
  alias: VH_Boden uebertrag
  description: 
  trigger:
  - platform: state
    entity_id:
    - sensor.gw2000a_v2_2_3_soil_moisture_1
  condition: []
  action:
  - service: homematicip_local.set_variable_value
    data:
      entry_id: XXXXXXX303e0d70d0d22f4bba694699a
      name: VH_Boden
      value: {{ states('sensor.gw2000a_v2_2_3_soil_moisture_1') | float | round(1) }}
  mode: single

**Reply by emufan** [reply]
Sven hat Dir ja auch eine fehlöerhafte Automatisierung geschickt. Fehler steht ja auch oben im Screenshot.

Fang nochmal oben an. YAML-Inhalt löschen und nicht an der Datei rumpfuschen. Neue leere Automatisierung im UI anlegen. Dann drei Punkte -> als YAML bearbeiten. Dann Fehler aus erste Automatisierung 

```
description:
```

korrigieren z.B. in

```
description: ""
```

Also z.B.

```
alias: VH_Boden uebertrag
description: 
trigger:
  - platform: state
    entity_id:
      - sensor.gw2000a_v2_2_3_soil_moisture_1
condition: []
action:
  - service: homematicip_local.set_variable_value
    data:
      entry_id: ca0c16aXXXXXXXXXXd22f4bba694699a
      name: VH_Boden VARIABLE IN HOMEMATIC
      value: |
        {{ states('sensor.gw2000a_v2_2_3_soil_moisture_1') | float(default=0) | round(1) }}
mode: single
```

Wobei halt die ID noch anzupassen ist.

Dann sollte es sich auch speichern lassen.

**Comment by Sven4597** [comment]


**Reply by Jonvelle53** [reply]
Hallo Sven,

habe als Anfänger eben diesen Eintrag von Dir für mein Problem weitergeleitet bekommen.
Leidewr kenne ich mich überhaupt nicht mit der Syntax aus.
Ich nehme an, dass ich Deine Zeilen oben nicht so ohne weiteres in meine config Yamml Datei übernehmen kann? Ich will Bodenfeuchtigkeitswerte aus Homeassistant in meine Homematic App übergeben. Also die Entitäten und S>ystemvariablen habe ich.
Aber wie gesagt, wie genau muss der Aufbau der Zeilen sein?

Achja - wo finde ich die Entry_ID der Systemvariable die ich beschreiben will?

Gruß
Jonvelle

**Reply by emufan** [reply]
> Achja - wo finde ich die Entry_ID der Systemvariable die ich beschreiben will?

z.B. Entwicklerwerkzeuge

Dort Service eingeben und bei Central im UI Deine CCU wählen. 

![image](https://github.com/danielperna84/hahomematic/assets/22775515/e83227ac-c705-4b82-bdfd-0fbdbae66e25)

Und danach dort in den YAML-Mode wechseln

![image](https://github.com/danielperna84/hahomematic/assets/22775515/b52ae84c-db2e-426d-ac88-2ef957784318)

**Reply by Jonvelle53** [reply]
Hallo EMUFan,

erstmal vielen Dank für Deine Geduld und schnelle kompetente Hilfe.
Es ist vollbracht. Alle Zustände meiner 10 Gartensprenger sowie deren Abhängigkeit und jeweiliger Bodenfeuchte kann ich nun sehr schön in meiner @Home App darstellen und steuern.

Es mag für einige Leute merkwürdig sein, dass ich Homeassistant nutze und dann Daten zurück zum Homematic System geben möchte, aber das liegt halt an meinen bescheidenen Programmier-Fähigkeiten.
Eine schöne und zweckmäßige Übersicht im HA zu programmieren braucht schon ne Menge Kenntnisse, die ich mir erst langsam aneignen muß.
Außerdem ist es für meine Frau vorerst auch einfacher unsere alte App zu nutzen.
Die Dashboards meines HA sind einfach zu voll und nicht sehr intuitiv zu bedienen.
Aber ich wollte halt alles unter einen Hut bringen und mich nicht von einzelnen Herstellern abhängig machen - Philips HUE (Beleuchtung) ZigBee, Homematic (Alarmanlage, Rauchmelder und Magnetventil-Steuerung), Shelly (Strom- und Leistungsüberwachung) sowie andere Zigbee und und und Anbieter.
Als ehemaliger schon seid 5 Jahren im Ruhestand befindlicher Außendienst Ing. kann ich mich ja nun austoben, und Dank solch großartiger Helfer hier im Netz wie Dir, komme ich dabei ganz gut voran. 
Also vielen Dank nochmal, bleib gesund und alles Gute für Dich.
Aus dem Spargelland Beelitz Grüßt Jonvelle

**Reply by Jonvelle53** [reply]
![image](https://github.com/danielperna84/hahomematic/assets/124449225/890da20b-86e2-4a3f-bc17-2b3797cffecd)

**Reply by emufan** [reply]
Nee nee, der Dank gebührt schon Sven als Ersthelfer. Ich hatte nur die Fehlermeldung und den Grund gesehen und bevor ihr hier weiter an YAML-Datei rumfummelt und der Fehler aber dennoch drin ist, wollte ich nur schnell beisteuern. Ich denke Sven hat halt seine Automatisierung anonymisiert und dabei nicht gesehen, dass "leer" dort nicht erlaubt ist.

> Eine schöne und zweckmäßige Übersicht im HA zu programmieren braucht schon ne Menge Kenntnisse, die ich mir erst langsam aneignen muß.

Das mach auf jeden Fall. Da ist die Lernkurve schnell und wirkliche Kenntnisse sind da auch oft nicht erforderlich. Gerade wenn es nur um Dashboards geht wie in Deinem Screenshot.

**Comment by Dark-Tower-Coder** [comment]
Hallo @Sven4597 
Ich nutze den selben Dienst um Daten aus HA in HM-Variablen schreiben zu können.
Aber was mich ein wenig stutzig macht ist folgende Information aus der Dokumentation:

> The service homematicip_local.set_variable_value can, but no longer has to, be used to write system variables.

Was soll das bedeuten?
Ist es veraltet und es gibt eine Alternative zu diesem Dienst?
Gibt es weitere Möglichkeiten?

Gruß
Tower

**Reply by emufan** [reply]
Siehe oben in diesem Thread und in der Doku. Extended SVs.

**Comment by Baxxy13** [comment]
Da ich heute auch ein bisschen mit dem Thema ("Synchronisierung eines HA-Sensor-Wertes mit einer Zentralen-Systemvariable") experimentiert habe...
```yaml
alias: HA-VM-RAM-Usage in RM-SysVar_V3
description: schreibt die RAM Nutzung der HA-VM in eine RM-SysVar
trigger:
  - platform: state
    entity_id:
      - sensor.memory_use_percent
condition: []
action:
  - delay:
      hours: 0
      minutes: 0
      seconds: 2
      milliseconds: 0
  - service: homematicip_local.set_variable_value
    data:
      entry_id: 2f499a9e0f9b82d60030da632c6c8657
      name: Homeassistant-VM-118_RAM-Nutzung
      value: "{{ trigger.to_state.state }}"
mode: single
```
Der Sensorwert (Trigger) wird mit 2s Verzögerung in die SysVar geschrieben.

**Reply by Jonvelle53** [reply]
Wo kann ich eigentlich etwas zur Syntax und den Möglichkeiten der Programmierung nachlesen, bzw. etwas lernen? Wenn möglich für Dummies erklärt!

**Reply by Sven4597** [reply]
Es gibt ganz viele Videos auf Youtube

**Reply by emufan** [reply]
> Da ich heute auch ein bisschen mit dem Thema ("Synchronisierung eines HA-Sensor-Wertes mit einer Zentralen-Systemvariable") experimentiert habe...

Wieso nutzt ihr eigentlich überhaupt die `homematicip_local.set_variable_value` für SVs und nicht die extended SVs mit Standard `number.set_value`, usw.? Dann hättet ihr doch sogar einen passenden Entität-Typ und UI-Änderbarkeit ohne Automatisierungen, Helper, usw.?

> Der Sensorwert (Trigger) wird mit 2s Verzögerung in die SysVar geschrieben.

Auch hier würde ich mal probieren, was wie im Vergleich der Verzug mit den Extended SVs wäre.

**Reply by emufan** [reply]
> Wo kann ich eigentlich etwas zur Syntax und den Möglichkeiten der Programmierung nachlesen, bzw. etwas lernen? Wenn möglich für Dummies erklärt!

Am besten öffnest Du dafür einen separaten Thread. Hier schon sehr OT und eigentlich für diese Integration ebenfalls.

**Reply by Jonvelle53** [reply]
Ok., habe bei youtube einen Lehrgang gefunden.
Allerdings stosse ich ja schon bei Euren Abkürzungen auf Probleme.
"Einen neuen Thread öffnen", ist ja noch verständlich, aber wo genau, und ohne Gefahr zu laufen, die Fragen an der falschen Stelle zu stellen. Oder was heißt "OT" bei Dir im Text? - ein Schreibfehler für "oft"?

Vielleicht kannst Du mir nochmal an Hand eines Beispiels erklären wo und wie ich eine Frage stellen kann?
Bei mir in HA steht jetzt in den Einstellungen zwei Warnungen.
Hier würde ich jetzt gern wissen, wie ich die Namen ändern soll?
Weil, demnächst die wohl nicht mehr funktionieren (ab 2024.2.0).

Was ich verstanden habe, dass ich diese Frage z.B. eben hier nicht stellen sollte. Ok. sorry dafür:

"Einige MQTT-Entitäten haben einen Entitätsnamen, der mit dem Gerätenamen beginnt. Dies ist nicht zu erwarten. Um einen doppelten Namen zu vermeiden, wird als Work-Around das Präfix des Gerätenamens aus dem Entitätsnamen entfernt. Bitte informiere den Betreuer der Softwareanwendung, die die betroffenen Entitäten bereitstellt, um dieses Problem zu beheben.

Liste der betroffenen Entitäten:

sensor.vibration_doppelflugel_kaminzimmer_strength
binary_sensor.gaszahler_contact
binary_sensor.vibration_doppelflugel_kaminzimmer_vibration
sensor.vibration_doppelflugel_kaminzimmer_angle_y
sensor.vibration_doppelflugel_kaminzimmer_power_outage_count
binary_sensor.buroog_power_update_available
sensor.buroog_power_power
sensor.buroog_power_energy
sensor.vibration_doppelflugel_kaminzimmer_battery
binary_sensor.buroog_power_consumer_connected .........."

**Reply by emufan** [reply]
OT heißt OffTopic = Nix mit dem Thema zu tun. Wie eigentlich auch die ganzen "wie bediene ich HA"-Fragen, da es hier ja um die HM Integration zum HA und nicht um HA selbst geht. Aber wenn es Sukram und Daniel erlaubt, dann könnte man für solch "Sonstiges" evtl. einen weiteren Diskussionsbeitrag unter Diskussionen machen a la "Allgemeine HA Fragen". Das meinte ich mit "neuer Thread". Aber das müssen die beiden erlauben.

Besser wäre natürlich, wenn Du Dich in https://community.home-assistant.io tummelst, aber das ist halt englisch. 

Zu den Warnungen. Steht auch dort. Warte ab, Da fehlen noch Folgeupdates: https://community.home-assistant.io/t/2023-8-translated-services-events-and-wildcards/598463/2

**Comment by Baxxy13** [comment]
> Wieso nutzt ihr eigentlich überhaupt...

Mit extended SV's samt `number.set_value` funktioniert das natürlich auch. 
Wenn ich jetzt einen vorhandenen HA-Sensorwert mit der extended SV's per Automatisierung synchronisiere wird im Hintergrund (vermutlich) auch nur `homematicip_local.set_variable_value` ausgeführt um den aktualisierten Wert Richtung CCU zu schieben.
Außerdem muss der SysVar-Scan aktiv sein damit das funktioniert.

Bei `homematicip_local.set_variable_value` geht das auch mit deaktiviertem SysVar-Scan und die SysVar braucht nicht in HA aktiv sein.
Es gibt halt mehrere Wege, kann ja jeder machen wie er mag.

> Auch hier würde ich mal probieren, was wie im Vergleich der Verzug mit den Extended SVs wäre.

Das hast du missverstanden. Die 2s Verzögerung sind doch absichtlich in der Automatisierung drin, der Text diente nur als Beschreibung und als Hinweis das man so direkt den Trigger-Wert bekommt und nicht einen Sensor-Wert "holen" muss.
